### PR TITLE
Bio.Align cleanup

### PR DESCRIPTION
**Pull request by mdehoon**
The alignment file parsers in `Bio.Align` return an `AlignmentIterator` object with metadata stored in the `.metadata` attribute. Converting the `AlignmentIterator` object to a list of alignments by calling `list` on it will lose the metadata. Redefining `AlignmentIterator` as a subclass of `list` instead of an iterator is problematic, as some alignment files are very large (e.g. MAF files) and the pairwise aligner can return an astronomically large number of alignments.

The solution then is to let `AlignmentIterator` derive from `list`, but to postpone loading the alignments into the list until actually needed. This lets a newly created `AlignmentIterator` object act as an iterator as long as possible, but as a list when needed.

For consistency between the `AlignmentIterator` class and the `PairwiseAlignments` class, the following inheritance scheme is used:
```
                                        .- ParsedAlignments <- AlignmentIterator (in the file parser modules)
list <- Alignments <- LazyAlignments <- |
                                        .- PairedAlignments (in the pairwise alignment module)
```
To prevent circular imports, the alignment class definitions are implemented in a private module `Bio.Align._base`, and the pairwise alignment code in a private module `Bio.Align._pairwise`.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

**Comment by mdehoon** [comment]
Attn @peterjc  in case you are interested.

**Comment by peterjc** [comment]
Slightly off topic, but we can configure the pre-commit hook to run black automatically (currently the user has to do this manually), and then enable the pre-commit.ci service to do this on pull requests.

Worth trying, or would it disrupt your workflow if the remote branch gets updated unexpectedly?

File: Bio/Align/_pairwise.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
+# Copyright 2000, 2004 by Brad Chapman.
```
Comment (peterjc) [inline Bio/Align/_pairwise.py:1]: Although the classes here were in ``Bio/Align/__init__.py`` I don't think that copyright block applies.

Rather it looks like your name should be here - it was added in https://github.com/biopython/biopython/pull/1655

**Comment by peterjc** [comment]
This seems overly complicated versus the approach of just exposing an alignment iterator and leaving it up to the user to call ``list(...)`` if they want random access.

Do you have a strong motivating use case?

**Comment by MarkusPiotrowski** [comment]
> Slightly off topic, but we can configure the pre-commit hook to run black automatically (currently the user has to do this manually), and then enable the pre-commit.ci service to do this on pull requests.

Or we can ask @mdehoon nicely (again?) to run pre-commit locally.

**Comment by mdehoon** [comment]
> This seems overly complicated versus the approach of just exposing an alignment iterator and leaving it up to the user to call `list(...)` if they want random access.
> 
> Do you have a strong motivating use case?

If you call `list`, the metadata will be lost, and list objects don't accept attributes, so a user cannot stick the metadata back on.
Without the metadata, getting alignments by parsing a file and writing those alignments back to a file will break.

**Comment by mdehoon** [comment]
@MarkusPiotrowski 

> Or we can ask @mdehoon nicely (again?) to run pre-commit locally.

Sorry, I don't know what pre-commit is.

**Comment by peterjc** [comment]
We used to use flake8 directly as a git pre-commit hook, but they deprecated that. Nowadays the tool pre-commit is widely used for running assorted git pre-commit hooks including flake8, black, spelling checks etc.

See https://github.com/biopython/biopython/blob/master/CONTRIBUTING.rst which covers installing this.

**Comment by peterjc** [comment]
With Bio.AlignIO and Bio.SeqIO there was usually file-level no metadata common to all the MSA or records. The only exception which comes to mind is SFF, where being pragmatic the flow tables etc are simple attached to every record - and the writer throws an exception if they are inconsistent.

It sounds like you're finding file-level metadata more common with the pairwise alignment formats?

**Comment by mdehoon** [comment]
> It sounds like you're finding file-level metadata more common with the pairwise alignment formats?

Emboss, exonerate, HHsuite, MAF, Mauve, PSL, SAM, and tabular output from BLAST and FASTA all have metadata.

An alternative to this PR is to have two classes to store alignments: An iterator-like class and a list-like class, with `alignments[:]` (where `alignments` is iterator-like) returning a list-like alignments object. We'd probably also want an abstract base class for the two classes.

**Comment by peterjc** [comment]
Do you think that alternative plan with two alternative classes would be simpler, or much the same complexity?

**Comment by mdehoon** [comment]
> Do you think that alternative plan with two alternative classes would be simpler, or much the same complexity?

Less complexity, because we won't have to override the list methods (`append`, `extend`, etc.).
But it will be more confusing, because sometimes `alignments[0]` will work (if `alignments` is list-like), and sometimes it won't (if `alignments` is iterator-like).

**Comment by mdehoon** [comment]
> We used to use flake8 directly as a git pre-commit hook, but they deprecated that. Nowadays the tool pre-commit is widely used for running assorted git pre-commit hooks including flake8, black, spelling checks etc.
> 
> See https://github.com/biopython/biopython/blob/master/CONTRIBUTING.rst which covers installing this.

Thanks. It would be great if this information could be merged into the GitUsage page. I always use that page when I forget how to use git with biopython.

**Comment by peterjc** [comment]
The page https://biopython.org/wiki/GitUsage already covers this under the section "Setting up a coding-style checker" - a long way down the page I admit. Does it just need to be more prominent?

**Comment by mdehoon** [comment]
> The page https://biopython.org/wiki/GitUsage already covers this under the section "Setting up a coding-style checker" - a long way down the page I admit. Does it just need to be more prominent?

ideally, this information would be under "Forking Biopython with your GitHub account" as a series of commands to be copied and executed. Usually when I start making a PR, I just copy the commands `git clone git@github.com:yourusername/biopython.git` and 
`git remote add upstream https://github.com/biopython/biopython.git` from that section, create a new branch, and start hacking. Many new contributors (and some old contributors like myself) are not familiar with pre-commit hooks and just want some fool-proof commands to be copied and executed.

**Comment by mdehoon** [comment]
As the metadata are important mainly when writing the alignments to file, another alternative would be to define a future `write` function like this:
```python
def write(alignments, handle, format, metadata=None):
```
where `alignments` is either an `AlignmentIterator` object obtained by parsing a file, or a plain list of `Alignment` objects. Then, if `metadata=None`, `write` will use the `.metadata` attribute on `alignments` if present. If the `metadata` argument is not `None`, then any `.metadata` attribute on `alignments` is ignored, and the `metadata` argument is used instead.

**Comment by mdehoon** [comment]
@peterjc  I have removed the Alignments inheritance hierarchy from this PR. What is left over is some cleanup in the `Bio.Align` parsing code.

**Review by peterjc — APPROVED** [review]
We there a reason to change the repr from:

```
<Bio.Align.Alignment object (2 rows x 14 columns) at 0x10403d850>
```

to

```
<Alignment object (2 rows x 14 columns) at 0x10403d850>
```

**Comment by mdehoon** [comment]
> We there a reason to change the repr from:
> 
> ```
> <Bio.Align.Alignment object (2 rows x 14 columns) at 0x10403d850>
> ```
> 
> to
> 
> ```
> <Alignment object (2 rows x 14 columns) at 0x10403d850>
> ```

I was thinking that that was closer to the Python default, for example
```python
>>> from collections import defaultdict
>>> d = defaultdict()
>>> d
defaultdict(None, {})
>>> from array import array
>>> a = array('i', [4,5,6])
>>> a
array('i', [4, 5, 6])
```
but now I am also finding some that include the module name:
```python
>>> from calendar import Calendar
>>> Calendar()
<calendar.Calendar object at 0x1088766d0>
```
In the near future, we may want to move the `Alignment` implementation to a private submodule and import it from `Bio.Align.__init__`, as `Bio.Align.__init__` is getting big. Then it may be better to use `Alignment` instead of `Bio.Align._someprivatemodule.Alignment`.

If you have a strong preference from `Bio.Align.Alignment`, I can change it back to that.

**Comment by peterjc** [comment]
I was just thinking we'd have several different alignment objects in Biopython now, and was it clear enough - that's all.

**Comment by mdehoon** [comment]
@peterjc  Thanks!

**Pull request merged**